File: jupyter/Qtjupyter/python/module/position.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from module.asset import Asset
from module.market import Market
from module.trade import Trade
from module.closingOrder import ClosingOrder
from module.derivative import Derivative

class Position:
    def __init__(self,trade: Trade):
        self.underlying = trade.underlying
        self.volume =trade.volume
        self.trade = trade
        self.derivative = trade.derivative
        self.closed = False
    def plotReturn(self):
        t0= self.derivative.uValue
        size = np.abs(t0/10)
        t = np.arange(t0-size,t0+size, size/100)
        #add Axis
        plt.axhline(0, color='black')
        plt.axvline(t0, color='black')
        #add Graph
        plt.plot(t,(self.derivative.f(t)-self.derivative.price)*self.volume,"r")
        plt.plot(t0,(self.derivative.f(t0)-self.derivative.price)*self.volume,"gx")
        
        plt.show()
    def printProps(self,withPlot=True):
        otp=f"""
        {'-'*40}
        Asset: {self.underlying.id}
        Positions type: {self.derivative.name}
        Opening Date: {self.trade.date}
        Underlying value: {self.derivative.uValue}
        Position value: {self.value}
        Amount: {self.volume}
        Price paid: {self.price}
        Strike price: {self.strike}
        Absolute return: {self.absReturn}
        Percentage return: {self.pctReturn}
        Log return: {self.logReturn}
        Currently closed: {self.closed}
        {'-'*40}
        """
        print(otp)
        if withPlot:
            self.plotReturn()
        
    
    
    @property
    def value(self) -> float:
        return self.volume*self.derivative.value
    @property
    def price(self)-> float:
        return self.derivative.price*self.volume
    @property
    def strike(self)-> float:
        return self.derivative.strike
    @property
    def absReturn(self)-> float:
        return self.derivative.absReturn*self.volume
    @property
    def pctReturn(self)-> float:
        return self.derivative.pctReturn
    @property
    def logReturn(self) -> float:
        if self.value<=0:
            return np.NINF
        return np.log(self.value/self.price)
    
    def getClosingOrder(self):
        if self.closed:
            return
        return ClosingOrder(self)
    def close(self) -> None:
        self.close = True
        

File: src/Position.jl
#Positio 

export 
    Position,
    volume,
    value,
    uValue,
    requestToClose 



"""
```julia
A = Asset()
B = Buy(A, 10)
O = Order(B, 100)
T = Trade(O)
P = Position(T)
```

"""
mutable struct Position
    derivative::Derivative
    trade::Trade  
    closed::Bool  
    requestToClose::Bool
    function Position(T::Trade)
        self = new()    
        self.derivative = T.derivative
        self.trade = T
        self.closed = false
        self.requestToClose = false
        return self
    end
end

Base.show(io::IO, P::Position) = P.closed ? 
print(io, "A closed position of $(volume(P)) $(P.derivative.name) on $(P.derivative.underlying.ticker)") : 
    print(io, "A open position of $(volume(P)) $(P.derivative.name) on $(P.derivative.underlying.ticker)")



volume(P::Position) = P.trade.volume
value(P::Position) = P.trade.volume * value(P.derivative)
uValue(P::Position) = P.trade.volume * uValue(P.derivative)
price(P::Position) = P.trade.volume * P.derivative.price
absReturn(P::Position) = P.trade.volume * absReturn(P.derivative)
pctReturn(P::Position) = pctReturn(P.derivative)
logReturn(P::Position) = value(P)<=0 ? -Inf : log(value(P)/price(P))


"""
    plot(P::<Position)

Plot the payoutstructure of the underlying derivative.


```

"""
function plot(P::Position)
    plot(P.derivative)
end

"""

    Trade(P::Position)

A trade that is the result of closing a position.

```julia
A = Asset()
B = Buy(A, 10)
O = Order(B, 100)
T = Trade(O)
cT=Trade(P)
value(cT)
```

"""
function Trade(P::Position)
    self = Trade()
    self.derivative = P.derivative
    self.volume = -P.trade.volume
    return self
end

function requestToClose(P::Position)
    P.requestToClose = true
    return P
end

function close(P::Position)
    T = Trade(P)
    P.closed = true
    return T
end



